File: src/ros2_bridge/mappers/control_msgs/joint_controller_state.rs
//! Typed mapper for `control_msgs/msg/JointControllerState`.

use crate::ros2_bridge::mapper::TypedTopicMapper;

pub(crate) fn joint_controller_state_to_bus(msg: ros_env::control_msgs::msg::JointControllerState) -> crate::control_msgs::msg::v1::JointControllerState {
    crate::control_msgs::msg::v1::JointControllerState {
        header: Some(crate::ros2_bridge::mappers::std_msgs::header::header_to_bus(msg.header)),
        set_point: msg.set_point,
        process_value: msg.process_value,
        process_value_dot: msg.process_value_dot,
        error: msg.error,
        time_step: msg.time_step,
        command: msg.command,
        p: msg.p,
        i: msg.i,
        d: msg.d,
        i_clamp: msg.i_clamp,
        antiwindup: msg.antiwindup,
    }
}

pub(crate) fn joint_controller_state_to_ros(bus: crate::control_msgs::msg::v1::JointControllerState) -> ros_env::control_msgs::msg::JointControllerState {
    ros_env::control_msgs::msg::JointControllerState {
        header: crate::ros2_bridge::mappers::std_msgs::header::header_to_ros(bus.header.unwrap_or_default()),
        set_point: bus.set_point,
        process_value: bus.process_value,
        process_value_dot: bus.process_value_dot,
        error: bus.error,
        time_step: bus.time_step,
        command: bus.command,
        p: bus.p,
        i: bus.i,
        d: bus.d,
        i_clamp: bus.i_clamp,
        antiwindup: bus.antiwindup,
    }
}

#[derive(Clone, Copy, Debug, Default)]
pub struct ControlMsgsJointControllerStateMapper;

impl TypedTopicMapper for ControlMsgsJointControllerStateMapper {
    type Ros = ros_env::control_msgs::msg::JointControllerState;
    type Bus = crate::control_msgs::msg::v1::JointControllerState;

    fn ros_to_bus(&self, msg: Self::Ros) -> crate::errors::Result<Self::Bus> {
        Ok(joint_controller_state_to_bus(msg))
    }

    fn bus_to_ros(&self, msg: Self::Bus) -> crate::errors::Result<Self::Ros> {
        Ok(joint_controller_state_to_ros(msg))
    }
}

File: bindings/python/robot_bus/ros2_bridge/mappers/control_msgs/joint_controller_state.py
"""Generated mapper for `control_msgs/msg/JointControllerState`."""

from __future__ import annotations

from robot_bus.ros2_bridge.mappers import _convert
from robot_bus.ros2_bridge.mappers.std_msgs.header import header_to_bus, header_to_ros

_BusMsg = None


def _bus_cls():
    global _BusMsg
    if _BusMsg is None:
        from robot_bus.control_msgs.msg.v1 import JointControllerState as BusMsg

        _BusMsg = BusMsg
    return _BusMsg


def joint_controller_state_to_bus(msg):
    BusMsg = _bus_cls()
    bus = BusMsg()
    bus.header.CopyFrom(header_to_bus(msg.header))
    bus.set_point = msg.set_point
    bus.process_value = msg.process_value
    bus.process_value_dot = msg.process_value_dot
    bus.error = msg.error
    bus.time_step = msg.time_step
    bus.command = msg.command
    bus.p = msg.p
    bus.i = msg.i
    bus.d = msg.d
    bus.i_clamp = msg.i_clamp
    bus.antiwindup = msg.antiwindup
    return bus


def joint_controller_state_to_ros(bus):
    from control_msgs.msg import JointControllerState as RosMsg

    out = RosMsg()
    out.header = header_to_ros(bus.header)
    out.set_point = bus.set_point
    out.process_value = bus.process_value
    out.process_value_dot = bus.process_value_dot
    out.error = bus.error
    out.time_step = bus.time_step
    out.command = bus.command
    out.p = bus.p
    out.i = bus.i
    out.d = bus.d
    out.i_clamp = bus.i_clamp
    out.antiwindup = bus.antiwindup
    return out


class ControlMsgsJointControllerStateMapper:
    _ros_type = None

    def ros_msg_type(self):
        cls = type(self)
        if cls._ros_type is None:
            from control_msgs.msg import JointControllerState as RosMsg

            cls._ros_type = RosMsg
        return cls._ros_type

    def ros_to_bus(self, msg) -> bytes:
        return joint_controller_state_to_bus(msg).SerializeToString()

    def bus_to_ros(self, payload: bytes):
        BusMsg = _bus_cls()
        bus = BusMsg()
        bus.ParseFromString(payload)
        return joint_controller_state_to_ros(bus)
